### PR TITLE
ENH: Fix support for matplotlib's constrained_layout (#25261)

### LINKED ISSUE #25261: plotting: subplots_adjust prevents use of constrained_layout=True
#### Code Sample, a copy-pastable example if possible

```python
# Your code here
import matplotlib.pyplot as plt
import pandas as pd

fig, axes = plt.subplots(2, constrained_layout=True)
times = pd.date_range(start='now', periods=10)
pd.DataFrame({'a': np.arange(10)}, index=times).plot(style='x', ax=axes[0])
```
#### Problem description
Plotting time-series uses pandas-internally a subplots_adjust, but due to this I am unable to use the new matplotlib constrained_layout that would take care of these things automatically.

#### Expected Output
A good layout that respects my constrained_layout setting to plt.subplots()

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.7.1.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.24.1
pytest: 4.2.0
pip: 19.0.2
setuptools: 40.7.3
Cython: 0.29.5
numpy: 1.16.1
scipy: 1.2.0
pyarrow: None
xarray: 0.11.3
IPython: 7.1.1
sphinx: 1.8.4
patsy: 0.5.1
dateutil: 2.8.0
pytz: 2018.9
blosc: None
bottleneck: None
tables: 3.4.4
numexpr: 2.6.9
feather: None
matplotlib: 3.0.2
openpyxl: None
xlrd: 1.2.0
xlwt: 1.3.0
xlsxwriter: 1.1.3
lxml.etree: 4.3.1
bs4: None
html5lib: None
sqlalchemy: 1.2.17
pymysql: None
psycopg2: 2.7.7 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None
</details>

### PR DESCRIPTION
**Pull request by mdruiter**
- [x] closes #25261
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by mdruiter** [comment]
The failing tests seem totally unrelated!?

**Review by jreback — CHANGES_REQUESTED** [review]
pls add the test from the OP

File: pandas/plotting/_matplotlib/boxplot.py [hunk 1]
```diff
@@ -227,7 +227,8 @@ def _grouped_plot_by_column(
 
     byline = by[0] if len(by) == 1 else by
     fig.suptitle(f"Boxplot grouped by {byline}")
-    fig.subplots_adjust(bottom=0.15, top=0.9, left=0.1, right=0.9, wspace=0.2)
+    if not hasattr(fig, "get_constrained_layout") or not fig.get_constrained_layout():
```
Comment (jreback) [inline pandas/plotting/_matplotlib/boxplot.py:230]: make a function for this
maybe_adjust_figure or similar

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/plotting/_matplotlib/boxplot.py (hunk 1 above)
Comment (mdruiter) [reply pandas/plotting/_matplotlib/boxplot.py]: Done. Hope this is what you had in mind.

File: pandas/tests/plotting/frame/test_frame_subplots.py [hunk 2]
```diff
@@ -482,6 +482,19 @@ def test_subplots_sharex_false(self):
...
+        idx = date_range(start="now", periods=10)
+        df = DataFrame(np.random.rand(10, 3), index=idx)
+        kwargs = {}
+        if hasattr(self.plt.Figure, "get_constrained_layout"):
+            kwargs["constrained_layout"] = True
+        fig, axes = self.plt.subplots(2, **kwargs)
+        with tm.assert_produces_warning(None):
+            df.plot(ax=axes[0])
+            with tm.ensure_clean(return_filelike=True) as path:
+                self.plt.savefig(path)
```
Comment (jreback) [inline pandas/tests/plotting/frame/test_frame_subplots.py:496]: do we need to save this?

File: pandas/plotting/_matplotlib/tools.py [hunk 3]
```diff
@@ -23,13 +23,24 @@
...
 
 
+def do_adjust_figure(fig):
+    if not hasattr(fig, "get_constrained_layout"):
+        return False
+    return not fig.get_constrained_layout()
+
+
+def maybe_adjust_figure(fig, *args, **kwargs):
+    if do_adjust_figure(fig):
```
Comment (jreback) [inline pandas/plotting/_matplotlib/tools.py:33]: can you add a doc-string (explaining the purpose here) & type what you can (e.g. fig)

File: pandas/plotting/_matplotlib/tools.py [hunk 4]
```diff
@@ -23,13 +23,24 @@
     from matplotlib.table import Table
 
 
+def do_adjust_figure(fig):
```
Comment (jreback) [inline pandas/plotting/_matplotlib/tools.py:26]: can you add a doc-string & type

File: doc/source/whatsnew/v1.3.0.rst [hunk 5]
```diff
@@ -341,7 +341,7 @@ Plotting
 ^^^^^^^^
 
 - Bug in :func:`scatter_matrix` raising when 2d ``ax`` argument passed (:issue:`16253`)
--
+- Support matplotlib's ``constrained_layout=True`` (:issue:`25261`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.3.0.rst:344]: can you be slightly more verbose here. this allows *passing* this option right?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/plotting/frame/test_frame_subplots.py (hunk 2 above)
Comment (mdruiter) [reply pandas/tests/plotting/frame/test_frame_subplots.py:496]: The warning does not happen until plt.show() or plt.savefig(), that's why. The saved image is directly removed.

File: doc/source/whatsnew/v1.3.0.rst (hunk 5 above)
Comment (mdruiter) [reply doc/source/whatsnew/v1.3.0.rst:344]: The change just prevents pandas from doing incompatible things when a figure has constrained_layout enabled.

File: pandas/plotting/_matplotlib/tools.py (hunk 4 above)
Comment (mdruiter) [reply pandas/plotting/_matplotlib/tools.py:26]: Sure.

File: pandas/plotting/_matplotlib/tools.py (hunk 3 above)
Comment (mdruiter) [reply pandas/plotting/_matplotlib/tools.py:33]: Sure.

File: pandas/tests/plotting/frame/test_frame_subplots.py (hunk 2 above)
Comment (jreback) [reply pandas/tests/plotting/frame/test_frame_subplots.py:496]: ok great

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mdruiter very nice!

**Pull request merged**